GitHubCore.ps1

```diff
@@ -989,3 +1070,98 @@ function Get-MediaAcceptHeader
+        '.mpg'  = 'video/mpeg' # MPEG Video
+        '.mpkg' = 'application/vnd.apple.installer+xml' # Apple Installer Package
+        '.mkv'  = 'video/x-matroska' # Matroska Multimedia Container
+        '.odp'  = 'application/vnd.oasis.opendocument.presentation' # OpenDocument presentation document
+        '.ods'  = 'application/vnd.oasis.opendocument.spreadsheet' # OpenDocument spreadsheet document
+        '.odt'  = 'application/vnd.oasis.opendocument.text' # OpenDocument text document
+        '.oga'  = 'audio/ogg' # OGG audio
+        '.ogg'  = 'application/ogg' # OGG audio or video
+        '.ogv'  = 'video/ogg' # OGG video
+        '.ogx'  = 'application/ogg' # OGG
+        '.opus' = 'audio/opus' # Opus audio
+        '.otf'  = 'font/otf' # OpenType font
+        '.png'  = 'image/png' # Portable Network Graphics
+        '.pdf'  = 'application/pdf' # Adobe Portable Document Format (PDF)
+        '.php'  = 'application/x-httpd-php' # Hypertext Preprocessor (Personal Home Page)
+        '.ps1'  = 'text/plain' # PowerShell script file
+        '.psd1' = 'text/plain' # PowerShell module definition file
+        '.psm1' = 'text/plain' # PowerShell module file
+        '.ppt'  = 'application/vnd.ms-powerpoint' # Microsoft PowerPoint
+        '.pptx' = 'application/vnd.openxmlformats-officedocument.presentationml.presentation' # Microsoft PowerPoint (OpenXML)
+        '.rar'  = 'application/vnd.rar' # RAR archive
+        '.rtf'  = 'application/rtf' # Rich Text Format (RTF)
+        '.sh'   = 'application/x-sh' # Bourne shell script
+        '.svg'  = 'image/svg+xml' # Scalable Vector Graphics (SVG)
+        '.swf'  = 'application/x-shockwave-flash' # Small web format (SWF) or Adobe Flash document
+        '.tar'  = 'application/x-tar' # Tape Archive (TAR)
+        '.tif'  = 'image/tiff' # Tagged Image File Format (TIFF)
+        '.tiff' = 'image/tiff' # Tagged Image File Format (TIFF)
+        '.ts'   = 'video/mp2t' # MPEG transport stream
+        '.ttf'  = 'font/ttf' # TrueType Font
+        '.txt'  = 'text/plain' # Text (generally ASCII or ISO 8859-n)
+        '.vsd'  = 'application/vnd.visio' # Microsoft Visio
+        '.wav'  = 'audio/wav' # Waveform Audio Format
+        '.weba' = 'audio/webm' # WEBM audio
+        '.webm' = 'video/webm' # WEBM video
+        '.webp' = 'image/webp' # WEBP image
+        '.woff' = 'font/woff' # Web Open Font Format (WOFF)
+        '.woff2' = 'font/woff2' # Web Open Font Format (WOFF)
+        '.xhtml' = 'application/xhtml+xml' # XHTML
+        '.xls'  = 'application/vnd.ms-excel' # Microsoft Excel
+        '.xlsx' = 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' # Microsoft Excel (OpenXML)
+        '.xml'  = 'application/xml' # XML
+        '.xul'  = 'application/vnd.mozilla.xul+xml' # XUL
+        '.zip'  = 'application/zip' # ZIP archive
+    }
+ }.GetEnumerator() | ForEach-Object {
+     Set-Variable -Scope Script -Option ReadOnly -Name $_.Key -Value $_.Value
+ }
```

Comment:
* It's [used here](https://github.com/microsoft/PowerShellForGitHub/blob/0f9278bceab032324d530e0841a36f411f2fe59e/GitHubCore.ps1#L236-L244) to determine the `ContentType` based on the extension.
* I think you misread this code.  This _is_ as hashtable with key/value pairs.  There are only 3 variables defined in this section: `defaultJsonBodyContentType`, `defaultInFileContentType`, `extensionToContentType`.
* This last statement is interesting.  Do you have more info on it?  Only 3 variables are being set this way in order to ensure that they're treated as constants (a similar pattern is used for other constants in this module).  If this is a bad pattern though, I'd love to learn more about it in order to make an educated decision on what to do moving forward in similar situations.

Reply:
> I think you misread this code. This is as hashtable with key/value pairs

Specifically I mean the `$val.GetEnumerator() | % { Set-Variable -Name $_.Key -Value $_.Value }` part.

If I do something like `@{ x = 1; y = 2; z = 3 }.GetEnumerator() | % { Set-Variable -Name $_.Key -Value $_.Value }` that sets three variables, `$x`, `$y` and `$z`.

But here, it doesn't seem like `$.zip` (for example) is ever used. What's the purpose of creating variables from the key/value pairs?

> It's used here to determine the ContentType based on the extension.

Yeah, this is exactly how I'd use it. In the form of the dictionary/hashtable. The way you're using it seems ideal to me. It's just the variable creation part that I don't understand.

> This last statement is interesting. Do you have more info on it?

Unfortunately, I think like most points of optimisation in PowerShell, it's totally undocumented.

In a module with few/no tight loops, mostly relying on calls to cmdlets, and being I/O intensive, I don't think it's anything to worry about. I have a bias against `Set-Variable` mainly because it's hard to analyse.

But if you're interested in more details, the code lives [here](https://github.com/PowerShell/PowerShell/blob/80ac3cc7354b0953e614c468d27d24f7cd845ac7/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L276-L280). Basically when a script is parsed for the first time, PowerShell does analysis to see if it can optimise the way variables are handled. When any of the `*-Variable` cmdlets are used, it decides it can't know statically, both because the name could be dynamically created and also because parameter binding is expensive.

Reply:
I still think you're misreading the code.
Let's try a modified version of it:

```powershell
@{
    defaultJsonBodyContentType = 'application/json; charset=UTF-8'
    defaultInFileContentType = 'text/plain'

    # Compiled mostly from https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
    extensionToContentType = @{
        '.3gp'  = 'video/3gpp' # 3GPP audio/video container
        '.3g2'  = 'video/3gpp2' # 3GPP2 audio/video container
        '.7z'   = 'application/x-7z-compressed' # 7-zip archive
        '.zip'  = 'application/zip' # ZIP archive
    }
 }.GetEnumerator() | ForEach-Object {
     Write-Host "creating $($_.Key)"
     Set-Variable -Scope Script -Option ReadOnly -Name $_.Key -Value $_.Value
 }
```

The output of that is:

```
creating defaultInFileContentType
creating extensionToContentType
creating defaultJsonBodyContentType
```

It's looping through as you described with your `@{ x = 1; y = 2; z = 3 }` example and creating 3 script variables...it's just that the last variable is a script variable that is a `[Hashtable]`.  It never creates a `$.zip` var.  It creates an `$extensionToContentType` var that lets me to this:

```powershell
$extensionToContentType['.zip']
```

to get this:

```
application/zip
```

Again, the only reason things are being done this way is because there doesn't appear to be any other way to enforce a `const`/`readonly` var within script scope.

Reply:
Oh I see -- I misread that brace. Yeah you're quite right